src/github.com/getlantern/balancer/dialer.go

```diff
@@ -94,29 +98,52 @@ func (d *dialer) start() {
 					d.OnClose()
 				}
 				return
-			case t := <-d.errCh:
-				lastFailed = t
+			case <-d.errCh:
+				atomic.StoreInt32(&d.consecSuccesses, 0)
+				log.Tracef("Mark dialer %s as inactive, scheduling check", d.Label)
+				atomic.AddInt32(&d.consecFailures, 1)
 			case <-timer.C:
 				ok := d.Check()
 				if ok {
-					lastCheckSucceeded = time.Now()
-					timer.Reset(longDuration)
-					consecCheckFailures = 0
+					// Note: we don't calculate avgConnTime here as the time to
+					// check a dialer may be different from actual requests.
+					atomic.AddInt32(&d.consecSuccesses, 1)
+					atomic.StoreInt32(&d.consecFailures, 0)
 				} else {
-					consecCheckFailures += 1
+					d.errCh <- struct{}{}
 				}
 			}
 		}
 	}()
 }
 
+func (d *dialer) metrics() metrics {
+	return metrics{atomic.LoadInt64(&d.avgConnTime), atomic.LoadInt32(&d.consecSuccesses), atomic.LoadInt32(&d.consecFailures)}
```

Comment:
You could actually get rid of the avgConnTime, consecSuccesses and consecFailures variables and have just a single metrics instance, and that way you don't have to construct a new one every time metrics() is called.